pkg/controllers/inventory/endpoint_handler.go

```diff
@@ -0,0 +1,166 @@
+	service := &v1.Service{}
+	err := c.Client.Get(
+		context.TODO(),
+		types.NamespacedName{
+			Name:      endpoint.Name,
+			Namespace: endpoint.Namespace,
+		},
+		service,
+	)
+	if err != nil {
+		log.Error(err, "Failed to get Service", "Name", endpoint.Name, "Namespace", endpoint.Namespace)
+		return
+	}
+	c.triggerPod(service)
+	c.triggerService(service)
+}
+
+func (c *InventoryController) triggerService(service *v1.Service) {
+	key, _ := keyFunc(service)
+	log.V(1).Info("Adding Service key to inventory object queue", "Service key", key)
+	c.inventoryObjectQueue.Add(inventory.InventoryKey{InventoryType: inventory.ContainerApplication, ExternalId: string(service.UID), Key: key})
+}
+
+func (c *InventoryController) triggerPod(service *v1.Service) {
+	// Get endpoint's corresponding pods and enqueue them
+	podIds, _ := c.service.GetPodIDsFromEndpoint(service)
+	for _, podID := range podIds {
+		ctx := context.TODO()
+		pod, err := getPodByUID(ctx, c.Client, types.UID(podID), service.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get Pod by UID", "PodID", podID, "Namespace", service.Namespace)
+			continue
+		}
+
+		serviceUIDs, err := getServicesUIDByPodUID(ctx, c.Client, pod.UID, pod.Namespace)
+		if err != nil {
+			log.Error(err, "Failed to get services UIDs by pod UID", "Pod UID", pod.UID, "Pod Namespace", pod.Namespace)
+			continue
+		}
+
+		containerApplicationInstance := containerinventory.ContainerApplicationInstance{
+			DisplayName:             pod.Name,
+			ContainerApplicationIds: serviceUIDs,
+			ExternalId:              podID,
+		}
+		c.service.UpdatePendingAdd(podID, &containerApplicationInstance)
+
+		podKey, keyErr := keyFunc(pod)
+		if keyErr != nil {
+			log.Error(keyErr, "Failed to get key for Pod", "Pod", pod.Name, "Namespace", pod.Namespace)
+			continue
+		}
+		log.V(1).Info("Adding Endpoint ContainerApplicationInstance key to inventory object queue", "ContainerApplicationInstance ExternalId", podID, "ContainerApplicationInstance Key", podKey)
+		c.inventoryObjectQueue.Add(inventory.InventoryKey{InventoryType: inventory.ContainerApplicationInstance, ExternalId: podID, Key: podKey})
+	}
+}
+
+func getPodByUID(ctx context.Context, c client.Client, uid types.UID, namespace string) (*v1.Pod, error) {
+	podList := &v1.PodList{}
+	if err := c.List(ctx, podList, &client.ListOptions{
+		Namespace: namespace,
+	}); err != nil {
+		return nil, fmt.Errorf("failed to list pods in namespace %s: %v", namespace, err)
+	}
+
+	for _, pod := range podList.Items {
+		if pod.UID == uid {
+			return &pod, nil
+		}
+	}
+
+	return nil, fmt.Errorf("pod with UID %s not found in namespace %s", uid, namespace)
+}
+
+func getServicesUIDByPodUID(ctx context.Context, c client.Client, podUID types.UID, namespace string) ([]string, error) {
+	// One pod can have multiple services, so we need to find all services associated with the pod
```

Comment:
I think the code could do the job. 
but the logic may be simple.
how about get the applicationid from the current inventoryapplicationinstance and merge them with the current service UID? you could refer to the current antrea.

Reply:
fixed

Reply:
As discussed in chat, we decide not to change this part for real time update. We can revisit it if performance issues found in the future.